src/orderbook/orderbooks.py

```diff
@@ -33,53 +34,33 @@ class SimpleOrderBook:
     def __init__(self, bid: Stack, ask: Stack, kafka_client: KafkaClient):
         self.bid = bid
         self.ask = ask
+        self.lock = threading.Lock()
         self.kafka_client = kafka_client
 
         self._QUOTES_TOPIC = "orders.topic"
         self._ORDER_STATUS_TOPIC = "order.status.topic"
         self._PRICE_TOPIC = "order.last_price.topic"
 
     def match(self, order: dict):
-        if order["order_type"] == "buy":
-            self._match_buy(order)
-        elif order["order_type"] == "sell":
-            self._match_sell(order)
-        else:
-            raise ValueError("Invalid order type")
-
-    def _match_buy(self, order: dict):
-        while (
-            order["quantity"] > 0
-            and self.ask
-            and order["price"] >= self.ask.peek()["price"]  # type: ignore
-        ):
-            best_ask = self.ask.peek()
-            if best_ask:
-                min_trade_quantity = min(order["quantity"], best_ask["quantity"])
-                print(f"trade: Buy {min_trade_quantity} @ {best_ask['price']}")
-                order["quantity"] -= min_trade_quantity
-                best_ask["quantity"] -= min_trade_quantity
-                if best_ask["quantity"] == 0:
-                    self.ask.pop()
-        if order["quantity"] > 0:
-            self.bid.push(order)
-
-    def _match_sell(self, order: dict):
-        while (
-            order["quantity"] > 0
-            and self.bid
-            and order["price"] <= self.bid.peek()["price"]  # type: ignore
-        ):
-            best_bid = self.bid.peek()
-            if best_bid:
-                min_trade_quantity = min(order["quantity"], best_bid["quantity"])
-                print(f"trade: Sell {min_trade_quantity} @ {best_bid['price']}")
-                order["quantity"] -= min_trade_quantity
-                best_bid["quantity"] -= min_trade_quantity
-                if best_bid["quantity"] == 0:
-                    self.bid.pop()
-        if order["quantity"] > 0:
-            self.ask.push(order)
+        """
+        Executes trades by matching buy orders with the lowest-priced ask
+        and sell orders with the highest-priced bid.
+        """
+        with self.lock:
+            is_buy = order["price"] > 0
+            book, opposite_book = (self.ask, self.bid) if is_buy else (self.bid, self.ask)
+            comparator = lambda p, q: p >= q if is_buy else abs(p) <= q
+            action = "Buy" if is_buy else "Sell"
+            while order["quantity"] > 0 and book and comparator(order["price"], book.peek()["price"]):
```

Comment:
I don't like the peek() here, seems a not thread safe operation.
For now we thread lock both match & push to stacks, but most likely not safe

Reply:
ok, let me think how can we improve this point